compositor_pipeline/src/pipeline/input/mp4/mp4_file_reader.rs

```diff
@@ -0,0 +1,369 @@
+use std::{
+    io::{Read, Seek},
+    os::unix::fs::MetadataExt,
+    sync::{atomic::AtomicBool, Arc},
+    time::Duration,
+};
+
+use bytes::{Buf, Bytes, BytesMut};
+use compositor_render::InputId;
+use crossbeam_channel::{Receiver, SendTimeoutError, Sender};
+use mp4::Mp4Reader;
+use tracing::{debug, span, warn, Level};
+
+use crate::{
+    pipeline::{
+        decoder::{AacDecoderOptions, AudioDecoderOptions, VideoDecoderOptions},
+        structs::{EncodedChunk, EncodedChunkKind},
+        AudioCodec, VideoCodec,
+    },
+    queue::PipelineEvent,
+};
+
+use super::{Mp4Error, Mp4ReaderOptions};
+
+type ChunkReceiver = Receiver<PipelineEvent<EncodedChunk>>;
+
+pub struct Mp4FileReader<DecoderOptions> {
+    stop_thread: Arc<AtomicBool>,
+    fragment_sender: Option<Sender<PipelineEvent<Bytes>>>,
+    decoder_options: DecoderOptions,
+}
+
+struct TrackInfo<DecoderOptions, SampleUnpacker: FnMut(mp4::Mp4Sample) -> Bytes> {
+    sample_count: u32,
+    timescale: u32,
+    track_id: u32,
+    decoder_options: DecoderOptions,
+    sample_unpacker: SampleUnpacker,
+    chunk_kind: EncodedChunkKind,
+}
+
+impl Mp4FileReader<AudioDecoderOptions> {
+    pub fn new_audio(
```

Comment:
It's not specific to this function. Try to avoid making everything pub by default, other than options most stuff should have `pub(super)` or `pub(crate)`